### PR TITLE
Add prefix option

### PR DESCRIPTION
**Pull request by wader**

### CODE REVIEW

**Comment by mholt** [comment]
I need to understand better what this does. Does this rewrite the request URI? i.e. does it strip the prefix? Or is it a read-only thing? Does it append the prefix the filepath when accessing disk?

**Comment by wader** [comment]
It is stripped from the URI path before access files in the root path and it is prepended to paths returned by PROPFIND.

See https://github.com/golang/net/blob/master/webdav/webdav.go for details.

**Comment by mholt** [comment]
Ahh.

So, URI manipulation is typically the role of the `rewrite` handler: https://caddyserver.com/docs/modules/http.handlers.rewrite

In other words:

```
uri strip_prefix /prefix/goes/here
webdav
```

What do you think? Do you think the webdav plugin should be modifying the request?

**Comment by wader** [comment]
Yeap would make sense but I have problems getting it to work with webdav clients if I just strip but don't prepend the paths returned in PROPFIND. Trying to understand what the webdav spec says about it.

https://tools.ietf.org/html/rfc4918#section-8.3

> WebDAV only uses one form of relative reference in its extensions, the absolute path.

And the examples seems to indicate absolute path are needed.

If that is the case we need to pass the Prefix to `webdav.Handler` or modify add that functionally to it :( or do you have any ideas?

**Comment by mholt** [comment]
Okay, I see.

If stripping the path is internal-only and doesn't actually modify the `req.URL.Path` value, then it's totally fine to configure the handler to do it.

Ideally, the handler that needs the path to be changed will copy the path value and change that, like, `path := strings.TrimPrefix(req.URL.Path, "/prefix")`, as opposed to: `req.URL.Path = strings.TrimPrefix(req.URL.Path, "/prefix")` which is destructive.

File: webdav.go [hunk 1]
```diff
@@ -30,6 +30,10 @@ type WebDAV struct {
 	// Accepts placeholders.
 	Root string `json:"root,omitempty"`
 
+	// The prefix to prepend to paths.
+	// Accepts placeholders.
+	Prefix string `json:"prefix,omitempty"`
```
Comment (mholt) [inline webdav.go:35]: Can we improve the godoc on this property? It's a little ambiguous/vague about what it actually does. I thought it stripped a path, not added one...

Comment (wader) [reply webdav.go:35]: Yeap should be more clear. Maybe some other term i better also? BasePath? RequestPath? 😕

Comment (mholt) [reply webdav.go:35]: `PathPrefix` maybe? It's hard to decide without knowing really well what it actually does.

Comment (wader) [reply webdav.go:43]: Added more detailed description and example that i hope i more clear. Decided to keep the option name. Think it looks strange if we have `root` but then `path_prefix`, should root be `root_path` then?

**Comment by wader** [comment]
Good! are the field comments used to generate documentation somehow? got a bit concerned that my example will get mangeled and look weird somehow

**Comment by wader** [comment]
Do you have any pointer or ideas for how the auth things should work? I might have a look it at some point

**Comment by mholt** [comment]
Yeah, they are -- in fact we render them as markdown. The website doesn't yet support external plugin docs, but it will eventually, so we might want to keep that in mind.

I'll add you as a collaborator to this repo and then next we can see if @hacdias wants to adopt it back upstream.

**Comment by mholt** [comment]
@wader The auth plugins use the authentication handler: https://caddyserver.com/docs/modules/http.handlers.authentication -- right now it's just basicauth.

But @greenpau is making some good progress on developing out the auth plugins further.

**Comment by greenpau** [comment]
> Do you have any pointer or ideas for how the auth things should work?

@wader , there will be `login`-type (was thinking naming it `caddy-auth-login` of a plugin, where user can submit username/password and authenticate against some user store. Upon successful authentication, a user will be issues a JWT token.

For example, `caddy-auth-saml` redirect a user to Azure portal for authentication, then validates SAML assertion it gets back, and sets the following token:
https://github.com/greenpau/caddy-auth-saml#jwt-token

That's the end of  `login`-type of a plugin.

Subsequent requests get intercepted by another "authentication" plugin, which is more of "authorization". It checks the validity of passed credentials (e.g. tokens, headers (incl. cookies), and grants a user access to a resource.

I am in the middle of the development on it https://github.com/greenpau/caddy-auth-jwt 

Hope this clarifies.

**Comment by mholt** [comment]
Thanks @greenpau 

To clarify, more than just JWT will be available, e.g. basicauth will continue to be supported too.

**Comment by greenpau** [comment]
> To clarify, more than just JWT will be available,

@mholt , absolutely! :+1: I actually never done [SSPI/Kerberos Interoperability with GSSAPI](https://docs.microsoft.com/en-us/windows/win32/secauthn/sspi-kerberos-interoperability-with-gssapi) development. Excited to get to it!

**Pull request merged**